lib/semantics/check-declarations.cc

```diff
@@ -352,6 +281,251 @@ void CheckHelper::CheckValue(
+            }
+          }
+        }
+      } else if (!IsIntentInOut(symbol)) {  // C1586
+        messages_.Say(
+            "non-POINTER dummy argument of PURE subroutine must have INTENT() or VALUE attribute"_err_en_US);
+      }
+    }
+  }
+}
+
+void CheckHelper::CheckProcEntity(
+    const Symbol &symbol, const ProcEntityDetails &details) {
+  if (details.isDummy()) {
+    const Symbol *interface{details.interface().symbol()};
+    if (!symbol.attrs().test(Attr::INTRINSIC) &&
+        (symbol.attrs().test(Attr::ELEMENTAL) ||
+            (interface && !interface->attrs().test(Attr::INTRINSIC) &&
+                interface->attrs().test(Attr::ELEMENTAL)))) {
+      // There's no explicit constraint or "shall" that we can find in the
+      // standard for this check, but it seems to be implied in multiple
+      // sites, and ELEMENTAL non-intrinsic actual arguments *are*
+      // explicitly forbidden.  But we allow "PROCEDURE(SIN)::dummy"
+      // because it is explicitly legal to *pass* the specific intrinsic
+      // function SIN as an actual argument.
+      messages_.Say("A dummy procedure may not be ELEMENTAL"_err_en_US);
+    }
+  }
+}
+
+void CheckHelper::CheckDerivedType(
+    const Symbol &symbol, const DerivedTypeDetails &details) {
+  CHECK(symbol.scope());
+  CHECK(symbol.scope()->symbol() == &symbol);
+  CHECK(symbol.scope()->IsDerivedType());
+  if (symbol.attrs().test(Attr::ABSTRACT) &&
+      (symbol.attrs().test(Attr::BIND_C) || details.sequence())) {
+    messages_.Say("An ABSTRACT derived type must be extensible"_err_en_US);
+  }
+  if (const DeclTypeSpec * parent{FindParentTypeSpec(symbol)}) {
+    const DerivedTypeSpec *parentDerived{parent->AsDerived()};
+    if (!IsExtensibleType(parentDerived)) {
+      messages_.Say("The parent type is not extensible"_err_en_US);
+    }
+    if (!symbol.attrs().test(Attr::ABSTRACT) && parentDerived &&
+        parentDerived->typeSymbol().attrs().test(Attr::ABSTRACT)) {
+      ScopeComponentIterator components{*parentDerived};
+      for (const Symbol &component : components) {
+        if (component.attrs().test(Attr::DEFERRED)) {
+          if (symbol.scope()->FindComponent(component.name()) == &component) {
+            SayWithDeclaration(component,
+                "Non-ABSTRACT extension of ABSTRACT derived type '%s' lacks a binding for DEFERRED procedure '%s'"_err_en_US,
+                parentDerived->typeSymbol().name(), component.name());
+          }
+        }
+      }
+    }
+  }
+}
+
+void CheckHelper::CheckGeneric(
+    const Symbol &symbol, const GenericDetails &details) {
+  const SymbolVector &specifics{details.specificProcs()};
+  const auto &bindingNames{details.bindingNames()};
+  std::optional<std::vector<Procedure>> procs{Characterize(specifics)};
+  if (!procs) {
+    return;
+  }
+  bool ok{true};
+  if (details.kind().IsAssignment()) {
+    for (std::size_t i{0}; i < specifics.size(); ++i) {
+      auto save{messages_.SetLocation(bindingNames[i])};
```

Comment:
I don't understand this use of "save".  It's being initialized, but I don't see it used anywhere.  But I don't get a compilation warning.

Reply:
`save` is an instance of `common::Restorer`. When it goes out of scope it undoes what `SetLocation` did. That way you don't need separate save and restore operations here and run the risk of forgetting the restore.

I've changed the name of the variable to `restorer` to make its purpose clearer.